### ISSUE TITLE
Rule proposal: warn against unnecessary `Number` calls

### ISSUE THREAD

**Issue by kkmuffme** [typescript-eslint/typescript-eslint#6385]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Report unnecessary calls to Number() constructor when type is a number already

### Fail Cases

```typescript
https://www.typescriptlang.org/play?#code/FAYw9gdgzgLgBAMzGOBeOBtAugbmKSWOAIwEMAnNOAOQFcBbYgU3IApFkA6AGyYgHMYACzgBKPEA


const foo = [];

const bar = Number( foo.length );
```

### Pass Cases

```typescript
const replace = Number( 'me' );
```

**Comment by bradzacher** [comment]
TBH in modern code one should avoid `Number()` because it performs a weak coercion to `number`. For example `Number(null) === 0`, `Number(true) === 1`, `Number('Infinity') === Number.Infinity`, `Number('     ') === 0`.

Instead using `parseInt`/`parseFloat` is much preferred because they only accept strings and don't do weird things like coercing whitespace-only strings to `0`, or parsing `Infinity` to the global Infinity value - they are all around much stricter.
They also strictly typed to only accept strings - meaning TS will error if you attempt to do something like `parseInt(0)`.

I'm not 100% sold on this being something that's of great value given the above and given that [step 1 of the number coercion algorithm](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-tonumber) is "if number return" - so it will be super low cost to have useless `Number()`'s in the codebase.

Would love to hear more feedback from others (cc @JoshuaKGoldberg).

**Comment by bradzacher** [comment]
@kkmuffme you can't really have an unnecessary `parseInt`/`parseFloat` because they are typed to only accept `string`.
So the only way it could be unnecessary is if you pass a string literal or regex literal typed value that clearly would make the function return `NaN`.

**Comment by JoshuaKGoldberg** [comment]
Maybe a general-purpose rule for unnecessary `Boolean(...)`, `Number(...)`, `String(...)`, and similar? I'm mildly in favor for this, but not very passionate about it. 😄

**Comment by Josh-Cena** [comment]
On MDN, I'm always slightly in favor of `Number()`. `parseInt()` does not allow `0x`/`0b`/`0o` prefixes, and its behavior of automatically ignoring invalid suffixes may not be desirable. If JavaScript is going to implicitly coerce your string to a number, it uses the same algorithm as `Number()` anyway, so I think the consistency of using the same coercion algorithm is more desirable than debating whether `Number("")` should be `NaN` or `0`. (`BigInt("")` also returns `0n`—that's another argument for consistency.)

**Comment by bradzacher** [comment]
> parseInt() does not allow 0x/0b/0o prefixes

How often do you need to parse input that's possibly in any of the valid, fully-qualified bases?!?
Like - I get that it might be nice - but I don't know of a case where you would actually want that behaviour aside from some super niche dev tools.

Personally I can count the number of times I've written something other than `parseInt(str)` / `parseInt(str, 10)` on one hand.

**Comment by bradzacher** [comment]
It's been a month and this issue has garnered no additional reactions or comments from the community. As such I'm going to close and reject this from the plugin, as it does not appear to represent a problem the wider community is interested in.
